// mongoengine/lib/constants/bson.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Bson;
(function (Bson) {
    Bson[Bson["double"] = 1] = "double";
    Bson[Bson["string"] = 2] = "string";
    Bson[Bson["object"] = 3] = "object";
    Bson[Bson["array"] = 4] = "array";
    Bson[Bson["binData"] = 5] = "binData";
    Bson[Bson["undefined"] = 6] = "undefined";
    Bson[Bson["objectId"] = 7] = "objectId";
    Bson[Bson["bool"] = 8] = "bool";
    Bson[Bson["date"] = 9] = "date";
    Bson[Bson["null"] = 10] = "null";
    Bson[Bson["regex"] = 11] = "regex";
    Bson[Bson["dbPointer"] = 12] = "dbPointer";
    Bson[Bson["javascript"] = 13] = "javascript";
    Bson[Bson["symbol"] = 14] = "symbol";
    Bson[Bson["javascriptWithScope"] = 15] = "javascriptWithScope";
    Bson[Bson["int"] = 16] = "int";
    Bson[Bson["timestamp"] = 17] = "timestamp";
    Bson[Bson["long"] = 18] = "long";
    Bson[Bson["decimal"] = 19] = "decimal";
    Bson[Bson["minKey"] = -1] = "minKey";
    Bson[Bson["maxKey"] = 127] = "maxKey";
})(Bson = exports.Bson || (exports.Bson = {}));
